w3af/core/ui/api/resources/sessions.py

```diff
@@ -0,0 +1,343 @@
+        for i in opts }
+
+    long_desc = " ".join(w3af.plugins.get_quick_instance(plugin_type, plugin)
+                         .get_long_desc().split())
+    enabled = plugin in w3af.plugins.get_enabled_plugins(plugin_type)
+
+    return jsonify({ 'configuration': plugin_opts,
+                     'description': long_desc,
+                     'enabled' : enabled })
+
+
+@app.route('/sessions/<int:scan_id>/plugins/<string:plugin_type>/<string:plugin>/',
+           methods=['PATCH'])
+@requires_auth
+@check_session_exists
+@check_plugin_type_exists
+@check_plugin_exists
+def set_plugin_config(**kwargs):
+    """
+    Allows a user to modify a single element of plugin configuration by sending
+    a PATCH request.
+
+    :return: 404 if scan or plugin does not exist.
+             400 if the PATCH request is not in valid format.
+             422 if the PATCH request is in valid format, but settings cannot be
+             applied (eg, negative numbers where none are accepted).
+
+             Otherwise, a JSON object confirming success and echoing the
+             changed setting.
+    """
+
+    scan_id = kwargs['scan_id']
+    plugin_type = kwargs['plugin_type']
+    plugin = kwargs['plugin']
+
+    w3af = SCANS[scan_id].w3af_core
+
+    plugin_opts = (
+        w3af.plugins.get_plugin_options(plugin_type, plugin) or
+        w3af.plugins.get_plugin_inst(plugin_type, plugin).get_options()
+    )
+
+    opt_names = dict(request.get_json(force=True))
+
+    if ('enabled' in opt_names and len(opt_names) == 1):
+        with SET_OPTIONS_LOCK:
+            enable_or_disable_plugin(w3af,
+                                     plugin,
+                                     plugin_type,
+                                     enable=request.json['enabled'])
+            return jsonify({
+                'message': 'success',
+                'modified': request.json
+                })
+    try:
+        # Remove 'enabled' if included as we must set this separately
+        opt_names.pop('enabled')
+    except LookupError:
+        pass
+
+    for opt_name in opt_names:
+        try:
+            opt_value = request.json[opt_name]
+            opt_type = plugin_opts[opt_name].get_type()
+        except BaseFrameworkException:
+            return jsonify({
+                'code': '400',
+                'message': '%s is not a valid option for plugin %s' %
+                    (opt_name,
+                     plugin)
+                }), 400
+        try:
+            if opt_type.lower() == 'url_list':
+                plugin_opts[opt_name].set_value([URL(o) for o in opt_value])
+            else:
+                plugin_opts[opt_name].set_value(opt_value)
+        except (AttributeError, BaseFrameworkException):
+            return jsonify({
+                'code': '422',
+                'message': 'Invalid %s option %s' % (opt_type, opt_value)
+                }), 422
+
+    with SET_OPTIONS_LOCK:
+        if 'enabled' in request.json:
```

Comment:
I've added a comment to clarify this:

`     if ('enabled' in opt_names and len(opt_names) == 1):
+    # If the request only contains 'enabled', set it here, return and skip
+    # the additional logic.
+    # If the request contains 'enabled' alongside other options, we handle it
+    # in the same lock section as other options below.
+    #
+    # This is done to ensure that the PATCH request is always atomic.
         with SET_OPTIONS_LOCK:
             enable_or_disable_plugin(w3af,`

I think it's valid, as the PATCH request must be atomic, but perhaps there's a neater way to handle it?